### ISSUE TITLE
Opensnitch UI taking quite long to start up

### ISSUE THREAD

**Issue by Shadow505** [evilsocket/opensnitch#566]
Hi,

I have noticed that the Opensnitch UI takes quite long to start up. 

To debug this, I decided to run opensnitch-ui via the console and can see that it takes about 55 seconds alone to load the translations (`Loading translations: /usr/lib/python3.10/site-packages/opensnitch/i18n locale`)

Then it takes another ~10 seconds until it's connected to the node.

This is happening on version 1.5.0 and as far as I can remember, it also happened on 1.4.3.

Besides being annoying, this is also an issue because if unknown connections are made during this time (by that I mean connections that you have not whitelisted yet in opensnitch), it will effectively block them until they are either approved or denied automatically, depending in your opensnitch settings.

My OS is Arch Linux. Currently running on kernel 5.15.7.

Best regards

**Comment by gustavo-iniguez-goya** [comment]
What Desktop Environment are you using? It takes that long to show up when there's no systray.

**Comment by Shadow505** [comment]
KDE with the regular taskbar

**Comment by gustavo-iniguez-goya** [comment]
uh, I don't know how to debug that problem. There's nothing before this log `Loading translations: /usr/lib/python3.10/site-packages/opensnitch/i18n locale`

https://github.com/evilsocket/opensnitch/blob/91ff72be7b364fbc08e4de5323f9943243b0aaf3/ui/bin/opensnitch-ui#L54

maybe it's something related to python3.10 ? https://www.reddit.com/r/archlinux/comments/rf6c84/psa_python_310_is_in_core_rebuild_your_aur/hoe7c5i/

**Comment by Shadow505** [comment]
Sorry, I don't quite understand how you want me to interpret your reply.

If you're concern is that I might be using an outdated opensnitch package from the AUR, that's not the case, I have compiled the package.

If you meant something else, please elaborate.

Thanks

**Comment by gustavo-iniguez-goya** [comment]
I meant that there's no operation before `Loading translations: /usr/lib/python3.10/site-packages/opensnitch/i18n locale`
You said that it took 55s to load the translations, but maybe I understood you wrong.

Are you saving the events to disk or in memory? (Preferences -> Database) If you haven't changed that option events are saved in memory. Saving events to disk could explain the long delays and freezes.

**Comment by gustavo-iniguez-goya** [comment]
One reason that could cause this behaviour is if the DB is stored on disk instead of memory.
With a db of ~ 50MB, it takes some seconds (5-20) to load just the DB (without even launching the main window), and after that the ConnectionsTableView takes another extra seconds (5-20) to load and displayed the data.

I'll debug these problems to see if I can improve the behaviour.

**Comment by Shadow505** [comment]
Hi,

The UI is in the process of starting up since close to 25 minutes now (and consuming 100% of a 5GHz core). I checked the database size, it is 1,7GB large. That is definitely quite large, I do however refuse to believe that there isn't a bug in the code. The file is sitting on a 2TB Samsung 980 Pro NVMe SSD. This really should not take so long, And the fact that the process is sitting at 100% CPU Load pretty much proves that the bottleneck is not the hard drive. 

 I will keep it running for a little longer, just to see how long it actually takes to start - if it even does before I go to sleep in a couple of minutes. If it does not start, I will manually open the Database file and  remove event logs. 

Besides that, my intent was not to indefinitely store all the events, I don't recall that there is/was an option to limit this. Not sure if I missed it.

Still, the fact that is taking so long and consuming so much CPU power is a very clear sign that there is something wrong. One could say that under this circumstances, the UI is completely unusable if you have a database of that size because no one is actually going to wait that long.

Hopefully this can be resolved. If there is any additional info that I can provide to help out, just let me know :)

Thanks for your work!

**Comment by Shadow505** [comment]
I'm sorry to say, it's definitely broken. It has been running over an hour now (constantly sitting at 100%), I therefore just killed that process. 

Best regards

**Comment by Shadow505** [comment]
> Besides that, my intent was not to indefinitely store all the events, I don't recall that there is/was an option to limit this. Not sure if I missed it.

I just checked it again and there is indeed no option available to limit the amount of events/connections that are being stored which means that you will eventually run into the situation that you won't be able to start the UI again.

**Comment by Shadow505** [comment]
> Besides that, my intent was not to indefinitely store all the events, I don't recall that there is/was an option to limit this. Not sure if I missed it.

I just checked it again and there is indeed no option available to limit the amount of events/connections that are being stored which means that you will eventually run into the situation that you won't be able to start the UI again.

@gustavo-iniguez-goya: **My suggestion as a hotfix:** Add an option to store the most relevant data on disk (rules, nodes), while storing connections, hosts, etc in memory.  Just using the in-memory storage is no option because you will loose your rules each time you do a reboot. This should however really be just a hotfix because also storing the other data (connections, etc) on disk (in a limited amount - which is atm also no possible) might be necessary, that depends on the use-case of each user. 

Just to clarify: With "this should really just be a hotfix", I _don't_ mean that this functionality should be removed afterwards, it would be a useful feature anyways, but the issue itself should be fixed as well due to the above reasons.

Finally going to bed now :)

**Comment by gustavo-iniguez-goya** [comment]
> Just using the in-memory storage is no option because you will loose your rules each time you do a reboot. 

mmh, not really. The GUI doesn't store the rules, it just handles them, is the daemon who stores them in disk under `/etc/opensnitchd/rules`. For that, rules must be of Duration == Always.

If you apply rules of Duration "Until Restart", rules will be valid until the daemon is restarted.
If you close the GUI you'll lose rules of temporary Duration (not Until Restart and not Always), and when you relaunch it, the daemon will send all active rules: Until Restart and Always, besides those that remain active (Duration 1h, 30m, etc).

**Comment by gustavo-iniguez-goya** [comment]
Please, could you tell me if you're saving events to disk or in memory?

If you're saving events to a DB on disk:
- Could you change it to in-memory (Preferences->Database), and confirm if all these problems still reproduce? 
- Also, could you post the size of the DB?
- Last but not least, I added options to limit the number of events to save, if you could test them and give me some feedback..

There's a bug that causes a delay when loading the GUI when there're many events in the DB. I'll try fix that soon.

**Comment by luhkius** [comment]
I had the same issue, and switching to the in-memory database fixed the problem. My db file was 967MB

**Comment by gustavo-iniguez-goya** [comment]
Thank you for the feedback @luhkius . What opensnitch version are you using?

**Comment by gustavo-iniguez-goya** [comment]
hey @themighty1 , sorry for pinging you again, but there's a performance issue on the Events tab when using a DB from disk.
With 300k records, updateDistinctIfNeeded takes 13' to complete 
https://github.com/evilsocket/opensnitch/blob/bb253629e8cff2bbe99727c36c1692d2ae01eca0/ui/opensnitch/customwidgets/main.py#L245

I've switched ConnectionsTableModel by GenericTableModel and works fine. It's not as smooth as ConnectionsTableModel, but I think that we could use it as a temporal fix for the next release, unless there's an easy way to improve updateDistinctIfNeeded. 
Searches are not as smooth as before, but at least the GUI doesn't block.

https://github.com/evilsocket/opensnitch/blob/master/ui/opensnitch/dialogs/stats.py#L419
```
                  model=GenericTableModel("connections", [                        
                      self.COL_STR_TIME,                                          
                      self.COL_STR_NODE,                                          
                      self.COL_STR_ACTION,                                        
                      self.COL_STR_DESTINATION,                                   
                      self.COL_STR_PROTOCOL,                                      
                      self.COL_STR_PROCESS,                                       
                      self.COL_STR_RULE,                                          
                  ]),#ConnectionsTableModel(),  
```

**Comment by luhkius** [comment]
> Thank you for the feedback @luhkius . What opensnitch version are you using?

My package manager says 1.4.3-2, the titlebar in the UI says 1.4.2, the bottom-right corner of the UI says 1.4.1

**Comment by gustavo-iniguez-goya** [comment]
ok thank you @luhkius . This problem will be fixed with the next version v1.5.0 which is about to be released.

**Comment by gustavo-iniguez-goya** [comment]
hey, I've added the mentioned solution to workaround this problem while we find a way of not causing long delays on start up.
@Shadow505 could you test it with latest sources?

If someone else can test it, please drop a comment.

**Comment by gustavo-iniguez-goya** [comment]
Hey @Shadow505 @luhkius , I've released a new version that should fix all these issues:
https://github.com/evilsocket/opensnitch/releases/tag/v1.5.0-rc.2

If you could try it and let me know if it works better would be very helpful. Thank you!

**Comment by gustavo-iniguez-goya** [comment]
I believe that this issue has been resolved in v1.5.0rc2. If you still experience problems, please, open a new issue.
